var/spack/repos/builtin/packages/spectre/package.py

```diff
@@ -0,0 +1,189 @@
+    version('2021.12.15', sha256='4bfe9e27412e263ffdc6fcfcb84011f16d34a9fdd633ad7fc84a34c898f67e5c')
+
+    # Configuration variants
+    variant('build_type',
+            default='Release',
+            values=("Release", "Debug", "RelWithDebInfo"),
+            description='CMake build type')
+    variant('executables',
+            values=any_combination_of(
+                'SolvePoisson1D', 'SolvePoisson2D', 'SolvePoisson3D',
+                'SolveElasticity2D', 'SolveElasticity3D', 'SolveXcts',
+            ),
+            description="Executables to install")
+    variant('python', default=False, description="Build Python bindings")
+    variant('doc', default=False, description="Build documentation")
+    # TODO: support installation of executables with shared libs
+    # variant('shared',
+    #         default=False,
+    #         description="Build shared libraries instead of static")
+    variant('pch',
+            default=True,
+            description="Compile with pre-compiled headers")
+    variant('memory_allocator',
+            values=('system', 'jemalloc'),
+            multi=False,
+            default='system',
+            description="Which memory allocator to use")
+    variant('formaline',
+            default=True,
+            description=("Write the source tree into simulation output files "
+                         "to increase reproducibility of results"))
+    variant('profiling',
+            default=False,
+            description="Enable options to make profiling SpECTRE easier")
+
+    # Compiler support
+    conflicts('%gcc@:6')
+    conflicts('%clang@:7')
+    conflicts('%apple-clang@:10')
+
+    # Build dependencies
+    depends_on('cmake@3.12:3', type='build')
+    depends_on('ninja', type='build')
+    depends_on('python@2.7:3.10', type='build')
+
+    # Link dependencies
+    depends_on('charmpp@6.10.2:7.0')
+    depends_on('blaze@3.8')
+    depends_on('boost@1.60:1.77')
+    depends_on('brigand@master')
+    depends_on('gsl')
+    depends_on('hdf5@:1.12~mpi')
+    depends_on('jemalloc', when='memory_allocator=jemalloc')
+    depends_on('libsharp@1.0.0:1.0~mpi~openmp')
+    depends_on('libxsmm@1.16.1:1')
+    depends_on('blas')
+    depends_on('lapack')
+    depends_on('yaml-cpp@0.6.3:0.7')
+
+    # Test dependencies
+    depends_on('catch2@2.8:2', type='test')
+    depends_on('py-numpy@1.10:1', type='test')
+    depends_on('py-scipy@1.0.0:1', type='test')
+    depends_on('py-h5py@3.0.0:3~mpi', type='test')
+
+    # Python bindings
+    with when('+python'):
+        extends('python')
+        depends_on('python@3.7:3.10', type=('build', 'run'))
+        depends_on('py-pybind11@2.6:2', type='build')
+        depends_on('py-numpy@1.10:1', type=('build', 'run'))
+        depends_on('py-scipy@1.0.0:1', type=('build', 'run'))
+        depends_on('py-matplotlib@3.0.0:3', type=('build', 'run'))
+        depends_on('py-h5py@3.0.0:3~mpi', type=('build', 'run'))
```

Comment:
You are specifying upper version bounds for all dependencies. This is not customary in Spack, unless you explicitly know that your package breaks otherwise.

There are two major philosophy lineages for upper version bounds, conservative and explorative. Spack is explorative.

Reply:
Wow that sounds scary :D I'm sure the code will break often if we leave the upper version open, e.g., we often had to patch support for new versions of Boost. Are you sure it's better to let people run into these issues, than to constrain to safe versions and expand them when we test them?

Reply:
I'm undecided on this issue. In small projects, having open upper bounds is definitively a win. In larger, active projects, closed upper bounds seem better. Some package are notorious for causing problems (e.g. HDF5, changing the API in a patch release). For most packages, open upper bounds are benign.

As I mentioned, it seems best to follow what most other Spack packages are doing.

I would add comments to any upper bounds you introduce. This prevents others from needlessly spending time trying to increase an upper bound (that is there to prevent a known failure), and prevents others from dismissing your package (for an outdated conservative upper bound that could easily be increased). You could also inquire on Slack regarding this.

Reply:
sounds good 👍